crates/subspace-networking/src/behavior/custom_record_store.rs

```diff
@@ -460,3 +462,63 @@ impl<'a> Iterator for ParityDbRecordIterator<'a> {
         })
     }
 }
+
+/// Record storage decorator. It wraps the inner record storage and monitors items number.
+pub struct LimitedSizeRecordStorageWrapper<RC = MemoryRecordStorage> {
+    // Wrapped record storage implementation.
+    inner: RC,
+    // Maintains FIFO queue to limit total item number.
+    fifo_keys: CircularBuffer<Key>,
+}
+
+impl<RC: for<'a> RecordStorage<'a>> LimitedSizeRecordStorageWrapper<RC> {
+    pub fn new(record_store: RC, max_items_limit: usize) -> Self {
+        let mut fifo_keys = CircularBuffer::new(max_items_limit);
+
+        // Initial cache loading.
+        for rec in record_store.records() {
+            let _ = fifo_keys.add(rec.key.clone());
+        }
+
+        trace!(size = fifo_keys.size(), "Record cache loaded.");
+
+        Self {
+            inner: record_store,
+            fifo_keys,
+        }
+    }
+}
+
+impl<'a, RC: RecordStorage<'a>> RecordStorage<'a> for LimitedSizeRecordStorageWrapper<RC> {
+    type RecordsIter = RC::RecordsIter;
+
+    fn get(&'a self, key: &Key) -> Option<Cow<'_, Record>> {
+        self.inner.get(key)
+    }
+
+    fn put(&mut self, record: Record) -> store::Result<()> {
+        let record_key = record.key.clone();
+
+        self.inner.put(record)?;
+
+        let evicted_key = self.fifo_keys.add(record_key);
+
+        if let Ok(Some(key)) = evicted_key {
+            trace!(?key, "Record evicted from cache.");
+
+            self.inner.remove(&key);
+        }
+
+        Ok(())
+    }
```

Comment:
There are a few issues I see with this:
* it stores unnecessary info in RAM, potentially quite a bit of it
* it doesn't have any notion of proximity to identity, thus you can easily wipe any node with puts of any content whatsoever; even if it was validated, without proximity node will end up storing something it wasn't supposed to

Not sure if it needs to be fixed though, it all seems like a wasted effort anyway considering we're not going to use puts long-term due to numerous issues with them and will replace with announcements anyway, so I'm a bit puzzled why are we implementing it right now in record store if it is a temporary short-term solution.